### PR TITLE
BUG: core: fix ilp64 blas dot/vdot/... for strides > int32 max

### LINKED ISSUE #17111: blas_stride has the wrong return type with HAVE_BLAS_ILP64
The return type of `blas_stride` is `int`, but it contains a comparison to `MAX_INT64`:

https://github.com/numpy/numpy/blob/64619c55896dad1927afe78c0702a58abf0d038f/numpy/core/src/multiarray/common.h#L299-L317

### PR DESCRIPTION
**Pull request by pv**
Fix overlooked int cast when HAVE_BLAS_ILP64 is defined.
It was supposed to cast to CBLAS_INT, not int.
Also add a regression test.

Fixes #17111

### CODE REVIEW

File: numpy/core/tests/test_regression.py [hunk 1]
```diff
@@ -2501,3 +2501,15 @@ def test__array_interface__descr(self):
                            formats=[np.int64, np.int64]))
         descr = np.array((1, 1), dtype=dt).__array_interface__['descr']
         assert descr == [('', '|V8')]  # instead of [(b'', '|V8')]
+
+    @pytest.mark.skipif(sys.maxsize < 2 ** 31 + 1, reason='overflows 32-bit python')
+    @requires_memory(free_bytes=9e9)
```
Comment (charris) [inline numpy/core/tests/test_regression.py:2506]: My machine requires more than this for the allocation.

File: numpy/core/tests/test_regression.py [hunk 2]
```diff
@@ -2501,3 +2501,15 @@ def test__array_interface__descr(self):
...
+    def test_dot_big_stride(self):
+        # gh-17111, blas with stride bigger than int32
+        int32_max = np.iinfo(np.int32).max
+        n = np.dtype(np.float32).itemsize * (int32_max + 1)
```
Comment (charris) [inline numpy/core/tests/test_regression.py:2510]: I think you should leave out the multiplication by itemsize here.

File: numpy/core/src/multiarray/common.h [hunk 3]
```diff
@@ -295,8 +295,14 @@ npy_memchr(char * haystack, char needle,
 /*
  * Convert NumPy stride to BLAS stride. Returns 0 if conversion cannot be done
  * (BLAS won't handle negative or zero strides the way we want).
+ *
+ * Return type same as CBLAS_INT (from npy_cblas.h; but we don't #include it here)
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h:299]: Can we just move `blas_stride` to that file? Any caller including this file without also including `npy_cblas` probably also has the wrong return type.

File: numpy/core/tests/test_regression.py (hunk 2 above)
Comment (charris) [reply numpy/core/tests/test_regression.py:2510]: Maybe use `int32_max + 2` instead,

**Review by eric-wieser — APPROVED** [review]

**Pull request merged**